contrib/database/sql/internal/dsn.go

```diff
@@ -135,3 +138,165 @@ func parseSQLServerDSN(dsn string) (map[string]string, error) {
 	delete(meta, "password")
 	return meta, nil
 }
+
+// parseSafe behaves like url.Parse, but if parsing fails it returns an
+// error with any credential part of the URL already scrubbed.
+func parseSafe(raw string) (*url.URL, error) {
+	u, err := url.Parse(raw)
+	if err == nil {
+		return u, nil
+	}
+	// url.Parse always wraps the real problem in *url.Error.
+	var ue *url.Error
+	if !errors.As(err, &ue) {
+		return nil, err
+	}
+	return nil, sanitizeURLError(ue)
+}
+
+// sanitizeURLError returns a copy of e whose URL field is redacted.
+func sanitizeURLError(e *url.Error) *url.Error {
+	// Best-case: we can still parse enough to use URL.Redacted().
+	if parsed, perr := url.Parse(e.URL); perr == nil {
+		e.URL = parsed.Redacted()
+		return e
+	}
+
+	// Fallback: use the comprehensive sanitize() function for all password patterns.
+	e.URL = sanitize(e.URL)
+	return e
```

Comment:
As noted below, the fallback makes it possible to leak credentials, in particular since the URL is malformed here, there could be typos in strategic places that cause credentials to still be leaked. I'd rather use an opaque string here...

Reply:
👍 Makes sense. It was hard to get the regexes right for the test cases that I have came up with.

Having an opaque string to say indicate that this URL in this error is redacted is fine for me.